File: modulo1/src/ros2_ws/src/fulgor_py_examples/fulgor_py_examples/add_two_ints_client.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from example_interfaces.srv import AddTwoInts
from functools import partial

class AddTwoIntsClient(Node):
    def __init__(self):
        super().__init__("add_two_ints_client")
        #creamos el cliente:
        self.client_ = self.create_client(AddTwoInts,"add_two_ints")

    #creamos un método para llamar al servicio para enviar la solicitud:
    def call_add_two_ints(self, a, b):
        #verificamos que el servicio está en ejecución:
        while not self.client_.wait_for_service(1.0):
            self.get_logger().warn("Esperando Add Two Ints server...")

        #creamos una solicitud:
        request = AddTwoInts.Request()
        request.a = a
        request.b = b
        #enviamos la solicitud:
        future = self.client_.call_async(request)
        #como ya tenemos un spin en el main, no hay que hacer otro.
        #generamos un cb para cuando llegue la respuesta:
        #partial, nos permite agregar más args, sino no podríamos agregar el arg request. Esto es-
        #-para los obj Future
        future.add_done_callback(partial(self.callback_call_add_two_ints,request=request))

    def callback_call_add_two_ints (self, future, request):
        response = future.result()
        self.get_logger().info(f"{request.a} + {request.b} = {response.sum}")


def main (args=None):
    rclpy.init(args=args)
    node = AddTwoIntsClient()
    #las sig líneas llaman directamente al método para enviar solicitudes:
    node.call_add_two_ints(20,5)
    node.call_add_two_ints(2,23)
    node.call_add_two_ints(3,1)
    #mantenemos en ejecución:
    rclpy.spin(node)
    rclpy.shutdown()

if __name__ == "__main__":
    main()

File: modulo1/src/ros2_ws/src/fulgor_cpp_examples/src/add_two_ints_client.cpp
#include "rclcpp/rclcpp.hpp"
#include "example_interfaces/srv/add_two_ints.hpp"
//permite usar literales de tiempo, ej: 1s
using namespace std::chrono_literals;
using namespace std::placeholders;
class AddTwoIntsClientNode: public rclcpp::Node
{
    public:
    AddTwoIntsClientNode(): Node("add_two_ints_client")
    {
        //creación del cliente para el servicio "add_two_ints"
        client_ = this->create_client<example_interfaces::srv::AddTwoInts>("add_two_ints");
    }
    //método para llamar al servicio
    void callAddTwoInts(int a, int b)
    {
        while(!client_->wait_for_service(1s))
        {
            RCLCPP_WARN(this->get_logger(), "esperando el servidor..");
        }
        //creación de la request
        auto request = std::make_shared<example_interfaces::srv::AddTwoInts::Request>();
        request->a = a;
        request->b = b;
        //envío asincrono de la request. El cb maneja la respuesta.
        client_->async_send_request(request, std::bind(&AddTwoIntsClientNode::callbackAddTwoInts, this, _1));
    }
    private:
    //cb se ejecuta cuando llega la rta del servicio
    void callbackAddTwoInts(rclcpp::Client<example_interfaces::srv::AddTwoInts>::SharedFuture future)
    {
        //obtiene la rta del servidor
        auto response = future.get();
        RCLCPP_INFO(this->get_logger(), "suma: %d", (int)response->sum);
    }
    //ptr al cliente del servicio
    rclcpp::Client<example_interfaces::srv::AddTwoInts>::SharedPtr client_;
};
int main (int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<AddTwoIntsClientNode>();
    node->callAddTwoInts(10,5);
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}